src/matrices.jl

```diff
@@ -0,0 +1,211 @@
+"""
+    compute_ptdf(system::System; block_size, reference_bus_index) -> KeyedArray
+    compute_ptdf(buses::DataFrame, branches::DataFrame; block_size, reference_bus_index) -> KeyedArray
+
+Takes a system, or tabular data for that system, representing a `M` branch, `N` bus grid
+and returns the `M * N` DC-Power Transfer Distribution Factor (DC-PTDF) matrix of the network.
+
+For a ~15,000 bus system with aggregated borders, this is expected to take ~1 minute.
+
+# Keywords
+- `block_size=13_000`: Block size to be used when partitioning a big matrix for inversion.
+- `reference_bus=first(keys(buses))`: The name of the reference bus.
+
+# Output
+- `::KeyedArray`: The PTDF matrix; the axes contain the branch names and bus numbers.
+
+!!! note
+    The input data must have no isolated components or islands.
+"""
+function compute_ptdf(system::System; kwargs...)
+    return compute_ptdf(get_buses(system), get_branches(system); kwargs...)
+end
+
+function compute_ptdf(
+    buses::Buses,
+    branches::Branches;
+    block_size=13_000,
+    reference_bus=nothing,
+)
+    bus_names = collect(keys(buses))
+    reference_bus_index = _reference_bus(reference_bus, bus_names)
+
+    incid_matrix = _incidence(buses, branches)
+    n_branches, n_buses = size(incid_matrix)
+
+    # Remove column related to reference bus from incidence matrix
+    incid_matrix = incid_matrix[:, Not(reference_bus_index)]
+
+    B_fl_tilde = sparse(diagm(_series_susceptance(branches))) * incid_matrix
+    B_bus_tilde_inv = big_mat_inv(
+        Matrix(incid_matrix' * B_fl_tilde),
+        block_size=block_size
+    )
+    ptdf_matrix = B_fl_tilde * B_bus_tilde_inv
+
+    # Add reference bus column back, filled with zeros
+    @views ptdf_matrix = hcat(
+        ptdf_matrix[:, 1:(reference_bus_index - 1)],
+        zeros(n_branches),
+        ptdf_matrix[:, reference_bus_index:end],
+    )
+
+    return KeyedArray(ptdf_matrix, (collect(keys(branches)), bus_names))
+end
+
+function _reference_bus(reference_bus, bus_names)
+    reference_bus === nothing && return 1
+
+    idx = findfirst(==(reference_bus), bus_names)
+    idx === nothing && throw(ArgumentError("Reference bus '$reference_bus' not found."))
+    return idx
+end
+
+"""
+    _series_susceptance(branch_df) -> Vector{Float64}
+
+Calculates the susceptance of the elements in the branch DataFrame. The calculation is
```

Comment:
Does this method work on a `DataFrame`?  It looks like it's intended to work on a `Branches`.